### PR TITLE
Ignore bundled code

### LINKED ISSUE #3368: Ignore bundled code
We would currently analyze (and fail) on files such as 

https://github.com/SonarResearchScanning/odoo/blob/master/addons/web/static/lib/pdfjs/build/pdf.js 

we should revisit our detection of generated code and avoid such files automatically.

### PR DESCRIPTION
**Pull request by saberduck**
Fixes #3368

Bundled code is detected and filtered out automatically

### CODE REVIEW

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.regex.Pattern;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.io.input.BoundedReader;
+import org.sonar.api.batch.fs.InputFile;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+
+import static java.util.regex.Pattern.DOTALL;
+
+public class BundleAssessor implements Assessor {
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java:33]: Maybe rename this to clarify that it looks for generated JS code? I'm particularly not a fan of "Bundle" which is extremely generic.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/Assessor.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.filter;
+
+import java.util.function.Predicate;
+import org.sonar.api.batch.fs.InputFile;
+
+public interface Assessor extends Predicate<InputFile> {
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/Assessor.java:25]: I found this answer for defining such an empty interface: https://stackoverflow.com/a/1437231/3967660
Could we discuss it as I'm not getting it.

File: sonar-javascript-plugin/src/main/resources/static/documentation.md [hunk 3]
```diff
@@ -52,6 +52,15 @@ is desired, it can be configured by setting `sonar.javascript.exclusions` proper
 `sonar.javascript.exclusions=""`, or to comma separated list of paths to be excluded. This property will exclude only JavaScript/TypeScript files, while `sonar.exclusions` property will exclude all files. `sonar.exclusions` property should be 
 preferred to configure general exclusions for the project.
 
+### Detection of code bundles
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/main/resources/static/documentation.md:55]: ... of generated code?

File: sonar-javascript-plugin/src/main/resources/static/documentation.md [hunk 4]
```diff
@@ -52,6 +52,15 @@ is desired, it can be configured by setting `sonar.javascript.exclusions` proper
...
+
+The analyzer will attempt to detect bundled code or generated code. This means code that was automatically transformed
+and optimized with tools such as Webpack and similar. We consider generated code out of scope of the analysis since
+developers are not able to act upon the findings in such code. Whenever generated code is detected, the analysis will
+print a log message: once per the whole project on `INFO` level, and for each file on the `DEBUG` level. If you want to
+opt-in for analyzing the generated code or in case the detection is incorrect, you can disable it by setting
+`sonar.javascript.detectBundles=false`.
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/main/resources/static/documentation.md:62]: detectGenerated?

File: sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/filter/JavaScriptExclusionsFileFilterTest.java [hunk 5]
```diff
@@ -217,6 +222,40 @@ void should_exclude_only_jsts_files() throws Exception {
...
+    assertThat(filter.accept(inputFile)).isFalse();
+    var logs = logTester.logs(LoggerLevel.INFO);
+    assertThat(logs).contains("Some of the project files were automatically excluded because they looked like generated " +
+      "code. Enable debug logging to see which files were excluded. You can disable bundle detection by setting " +
+      "sonar.javascript.detectBundles=false");
+
+    assertThat(logTester.logs(LoggerLevel.DEBUG)).contains("File bootstrap.js was excluded because it looks like a " +
+      "bundle. (Disable detection with sonar.javascript.detectBundles=false)");
+  }
+
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/filter/JavaScriptExclusionsFileFilterTest.java:258]: Aren't we missing a test when exclusion is not set?

Comment (ilia-kebets-sonarsource) [reply sonar-javascript-plugin/src/test/java/org/sonar/plugins/javascript/filter/JavaScriptExclusionsFileFilterTest.java:258]: Why not separate these 2 cases?

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  private static final Logger LOG = Loggers.get(BundleAssessor.class);
+  static final String PROPERTY = "sonar.javascript.detectBundles";
+  private static final Pattern COMMENT_OPERATOR_FUNCTION = Pattern.compile("/\\*.*\\*/\\s*[!;+(]function ?(?: [_$a-zA-Z][_$a-zA-Z0-9]*)?\\(", DOTALL);
```
Comment (ilia-kebets-sonarsource) [inline sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java:37]: Could you link to what you based this regex on? or explain it?

**Review by ilia-kebets-sonarsource — CHANGES_REQUESTED** [review]
See my comment for minor changes. LMK if you wish to discuss it.

File: sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java (hunk 1 above)
Comment (saberduck) [reply sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/filter/BundleAssessor.java:33]: let's settle on bundle

**Review by ilia-kebets-sonarsource — APPROVED** [review]
LGTM, thanks for the changes after review!

**Pull request merged**